lens/fetcher.go

```diff
@@ -0,0 +1,397 @@
+// Copyright 2023 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package lens
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/fxamacker/cbor/v2"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/client/request"
+	"github.com/sourcenetwork/defradb/core"
+	"github.com/sourcenetwork/defradb/datastore"
+	"github.com/sourcenetwork/defradb/db/fetcher"
+	"github.com/sourcenetwork/defradb/planner/mapper"
+)
+
+// todo: The code in here can be significantly simplified with:
+// https://github.com/sourcenetwork/defradb/issues/1589
+
+type lensedFetcher struct {
+	source   fetcher.Fetcher
+	registry client.LensRegistry
+	lens     Lens
+
+	txn datastore.Txn
+
+	col *client.CollectionDescription
+	// Cache the fieldDescriptions mapped by name to allow for cheaper access within the fetcher loop
+	fieldDescriptionsByName map[string]client.FieldDescription
+
+	targetVersionID string
+	// We cache the target schema version ID as bytes so we dont have to bother
+	// casting it within the fetch loop.
+	targetVersionIDBytes []byte
```

Comment:
suggestion: This feels unnecessary, I would pick one or the other, whatever tiny perf we might gain, I don't think is worth the added confusion/duplication.

Reply:
That is fair enough, I think this largely stems from my allergy to doing unnecessary stuff in loops instead of any real tangible benefit, and the maintenance cost is non-zero. I'll remove it.

- [ ] remove one `targetVersionID` prop and just cast in loop